apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/R4ClaimResponseResourceProviderTest.java

```diff
@@ -0,0 +1,257 @@
+  }
+
+  @Test
+  public void shouldHaveReadMethodAnnotatedWithRead() throws NoSuchMethodException {
+    R4ClaimResponseResourceProvider provider = new R4ClaimResponseResourceProvider();
+
+    Method readMethod = provider.getClass().getMethod("read", IdType.class, RequestDetails.class);
+
+    Read annotation = readMethod.getAnnotation(Read.class);
+
+    assertNotNull(annotation);
+  }
+
+  @Test
+  public void shouldHaveReadMethodWithAnnotatedId() throws NoSuchMethodException {
+    R4ClaimResponseResourceProvider provider = new R4ClaimResponseResourceProvider();
+
+    Method readMethod = provider.getClass().getMethod("read", IdType.class, RequestDetails.class);
+
+    IdParam annotation = null;
+
+    for (Parameter p : readMethod.getParameters()) {
+      if (p.getType() == IdType.class) {
+        annotation = p.getAnnotation(IdParam.class);
+      }
+    }
+
+    assertNotNull(annotation);
+  }
+
+  @Test
+  public void shouldThrowIllegalArgumentExceptionIfIdIsNull() {
+    Exception expected = new IllegalArgumentException("Resource ID can not be null");
+    Exception actual =
+        AssertUtils.catchExceptions(() -> new R4ClaimResponseResourceProvider().read(null, null));
+
+    AssertUtils.assertThrowEquals(expected, actual);
+  }
+
+  @Test
+  public void shouldThrowIllegalArgumentExceptionIfIdVersionIsNotNumeric() {
+    IdType id = new IdType(null, "f-123", null, "null");
+
+    Exception expected = new NumberFormatException("For input string: \"null\"");
+    Exception actual =
+        AssertUtils.catchExceptions(() -> new R4ClaimResponseResourceProvider().read(id, null));
+
+    AssertUtils.assertThrowEquals(expected, actual);
+  }
+
+  @Test
+  public void shouldThrowIllegalArgumentExceptionIfIdVersionIsNumeric() {
+    IdType id = new IdType(null, "f-123", null, "123");
+
+    Exception expected = new IllegalArgumentException("Resource ID must not define a version.");
+    Exception actual =
+        AssertUtils.catchExceptions(() -> new R4ClaimResponseResourceProvider().read(id, null));
+
+    AssertUtils.assertThrowEquals(expected, actual);
+  }
+
+  @Test
+  public void shouldThrowIllegalArgumentExceptionIfIdPartIsNull() {
+    IdType id = new IdType(null, null, null, null);
+
+    Exception expected = new IllegalArgumentException("Resource ID can not be null/blank");
+    Exception actual =
+        AssertUtils.catchExceptions(() -> new R4ClaimResponseResourceProvider().read(id, null));
+
+    AssertUtils.assertThrowEquals(expected, actual);
+  }
+
+  @Test
+  public void shouldThrowIllegalArgumentExceptionIfIdPartIsBlank() {
+    IdType id = new IdType(null, null, " ", null);
+
+    Exception expected = new IllegalArgumentException("Resource ID can not be null/blank");
+    Exception actual =
+        AssertUtils.catchExceptions(() -> new R4ClaimResponseResourceProvider().read(id, null));
+
+    AssertUtils.assertThrowEquals(expected, actual);
```

Comment:
These exception tests are good.  They are pretty repetitive though.  Maybe add a method in AssertUtils for the pattern?  (i.e. takes expected and Callable as params).

Reply:
We could, though I think these tests are pretty small already.  In the end, passing the expected and callable as parameters to another testing utility method wouldn't be any more terse then if I just modified the current test to have the `expected` and `actual` values inline in the `assertThrowEquals()` invocation.  I thought about parameterized tests, but the built in JUnit one isn't ideal, as it runs every test for every parameter if I remember correctly.  There are solutions to get around this like using an enclosed test runner, but that ends up getting rather messy looking adding extra layers of complexity to the test.  There are also some pretty solid parameterized test libraries out there that work a bit better, but I didn't think it was worth pulling in another library for this at the moment.